src/server/src/services/github.js

```diff
@@ -31,27 +31,43 @@ const OctokitWithPluginsAndDefaults = Octokit.plugin(
     }
 })
 
-async function callGithub(octokit, obj, fun, arg, cacheKey, cacheTime) {
-    if (cacheKey && !config.server.nocache) {
-        const cachedRes = cache.get(cacheKey)
-        if (cachedRes) {
-            logger.info(`Result returned from cache for ${obj}.${fun}`)
-            return cachedRes
+async function callGithub(octokit, obj, fun, arg, cacheKey, isUseETag) {
+    let res
+
+    if (isUseETag && !config.server.nocache) {
+        const cachedData = cache.get(cacheKey)
+        if (cachedData) {
+            arg.headers = {
+                'If-None-Match': stringify(cachedData.headers.etag.split('"')[1])
+            }
+            delete arg.isUseETag
         }
     }
-    let res
-    if (fun.match(/list.*/g)) {
-        const options = octokit[obj][fun].endpoint.merge(arg)
-        res = {
-            data: await octokit.paginate(options)
+
+    try {
+        if (fun.match(/list.*/g)) {
+            const options = octokit[obj][fun].endpoint.merge(arg)
+            res = {
+                data: await octokit.paginate(options)
+            }
+        } else {
+            res = await octokit[obj][fun](arg)
+        }
+    } catch (error) {
+        if (error.status === 304) {
+            logger.info(`Conditional request using etag for ${obj}.${fun}`)
+            const cachedData = cache.get(cacheKey)
+            if (cachedData) {
+                return cachedData
+            }
+        } else {
+            logger.error(new Error(error).stack)
+            throw new Error(`${fun}.${obj}: ${error}`)
         }
-    } else {
-        res = await octokit[obj][fun](arg)
-        // logger.info(`Result for ${obj}.${fun}`)
     }
 
-    if (res && cacheTime) {
-        cache.put(cacheKey, res, 1000 * cacheTime)
+    if (isUseETag && res && res.headers && res.headers.etag) {
+        cache.put(cacheKey, res, 1000 * 60 * 60) // caching for one hour - rate limit will be reset
```

Comment:
The cache time shouldn't have a limit. Due to the etag we will anyway get a response from GitHub if anything changes and although the rate limit resets every hour with several thousand repositories we can run into the rate limit anyway.